### PR TITLE
Add support for Wav2Vec2

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR adds optimized support for Wav2Vec2 for the following tasks:
- audio classification
- speech recognition

A multi-card regression test has been added for audio classification and the documentation has been updated.
For speech recognition, the regression test is currently commented because it is a bit too long. This will be refined in another PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/audio-classification/README.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    --dataloader_num_workers 8 \
+    --seed 0 \
+    --use_habana \
+    --use_lazy_mode \
+    --gaudi_config_name Habana/wav2vec2 \
+    --throughput_warmup_steps 2
+```
+
+On 8 HPUs, this script should run in ~12 minutes and yield accuracy of **80.38%**.
+
```
Comment (libinta) [inline examples/audio-classification/README.md:92]: seems this is not good scaling? do we know why?
should we actually put some notes as the model is not fully optimized, the optimization can be done through gaudi_config and training parameter?

**Review by libinta — COMMENTED** [review]
for models that are not fully optimized, should we mention about it?

File: examples/audio-classification/README.md (hunk 1 above)
Comment (regisss) [reply examples/audio-classification/README.md:92]: What do you mean? The multi-HPU example uses a different dataset and different hyperparameters from the single-HPU one.

File: examples/audio-classification/README.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+On a single HPU, this script should run in ~13 minutes and yield accuracy of **97.96%**.
+
+> If your model classification head dimensions do not fit the number of labels in the dataset, you can specify `--ignore_mismatched_sizes` to adapt it.
+
+
+## Multi-HPU
+
+The following command shows how to fine-tune [wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base) for 🌎 **Language Identification** on the [CommonLanguage dataset](https://huggingface.co/datasets/anton-l/common_language) on 8 HPUs.
```
Comment (libinta) [inline examples/audio-classification/README.md:59]: should it be "Audio Classification" instead of "Language Identification"?

File: examples/speech-recognition/README.md [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    --throughput_warmup_steps="2"
+```
+
+On a single HPU, this script should run in *ca.* 6 hours and yield a CTC loss of **0.059** and word error rate of **0.0423**.
+
+> If your data has a sampling rate which is different from the one of the data the model was trained on, this script will raise an error.
+> Resampling with the `datasets` library is not supported on HPUs yet.
+
+### Multi-GPU CTC
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:88]: change to Multi-HPU

File: examples/speech-recognition/README.md [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+```
+
+[The documentation](https://huggingface.co/docs/optimum/habana/usage_guides/deepspeed) provides more information about how to use DeepSpeed within Optimum Habana.
+
+> If your data has a sampling rate which is different from the one of the data the model was trained on, this script will raise an error.
+> Resampling with the `datasets` library is not supported on HPUs yet.
+
+
+<!-- ### Multi GPU CTC with Dataset Streaming
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:173]: Multi-HPU

File: examples/speech-recognition/README.md [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+	--chars_to_ignore , ? . ! - \; \: \" “ % ‘ ” � \
+	--max_duration_in_seconds="20" \
+	--shuffle_buffer_size="500" \
+	--fp16 \
+	--push_to_hub \
+	--do_train --do_eval \
+	--gradient_checkpointing**
+```
+
+On 4 V100 GPUs, this script should run in *ca.* 3h 31min and yield a CTC loss of **0.35** and word error rate
```
Comment (libinta) [inline examples/speech-recognition/README.md:223]: should we  remove GPU info here?

File: examples/speech-recognition/README.md [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+```bash
+OMP_NUM_THREADS=1 python run_speech_recognition_ctc ...
+```
+
+If the environment variable is not set, the training script might freeze, *i.e.* see: https://github.com/pytorch/audio/issues/1021#issuecomment-726915239
+
+---
+
+### Single GPU Seq2Seq
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:318]: change to HPU

File: examples/speech-recognition/README.md [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+The following command shows how to fine-tune [XLSR-Wav2Vec2](https://huggingface.co/transformers/main/model_doc/xlsr_wav2vec2.html) on [Common Voice](https://huggingface.co/datasets/common_voice) using a single GPU in half-precision.
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:320]: Should we do global replacement of GPU with HPU?

File: examples/speech-recognition/README.md [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+	--generation_num_beams="1" \
+	--do_train --do_eval \
+	--do_lower_case
+```
+
+On a single V100 GPU, this script should run in *ca.* 5 hours and yield a
+cross-entropy loss of **0.405** and word error rate of **0.0728**.
+
+### Multi GPU Seq2Seq
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:360]: same here

File: examples/speech-recognition/README.md [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+	--gradient_checkpointing \
+	--fp16 \
+	--group_by_length \
+	--predict_with_generate \
+	--do_train --do_eval \
+	--do_lower_case
+```
+
+On 8 V100 GPUs, this script should run in *ca.* 45 minutes and yield a cross-entropy loss of **0.405** and word error rate of **0.0728**
+
```
Comment (libinta) [inline examples/speech-recognition/README.md:397]: should we put HPU results here?

File: optimum/habana/modeling_utils.py [hunk 10]
```diff
@@ -38,6 +47,12 @@ def adapt_transformers_to_gaudi(use_habana_mixed_precision: bool):
     # Optimization tweak for ViT
     ViTSelfAttention.forward = gaudi_vit_self_attention_forward
 
+    # Optimization tweak for Wav2Vec2
+    _compute_mask_indices = _gaudi_wav2vec2_compute_mask_indices
+    _sample_negative_indices = _gaudi_wav2vec2_sample_negative_indices
+    Wav2Vec2Model._mask_hidden_states = _gaudi_wav2vec2_mask_hidden_states
+    Wav2Vec2Model.forward = gaudi_wav2vec2_forward
+
```
Comment (libinta) [inline optimum/habana/modeling_utils.py:55]: why do we need gaudi specific change?

File: examples/audio-classification/README.md (hunk 2 above)
Comment (regisss) [reply examples/audio-classification/README.md:59]: Language identification is a specific type of audio classification so I think this is fine.

File: examples/speech-recognition/README.md (hunk 3 above)
Comment (regisss) [reply examples/speech-recognition/README.md:88]: Good catch!

File: examples/speech-recognition/README.md (hunk 4 above)
Comment (regisss) [reply examples/speech-recognition/README.md:173]: This is commented, I'm just going to remove it for now.

File: examples/speech-recognition/README.md (hunk 5 above)
Comment (regisss) [reply examples/speech-recognition/README.md:223]: Same

File: examples/speech-recognition/README.md (hunk 6 above)
Comment (regisss) [reply examples/speech-recognition/README.md:318]: Same

File: optimum/habana/modeling_utils.py (hunk 10 above)
Comment (regisss) [reply optimum/habana/modeling_utils.py:55]: To generate mask and negative indices directly on HPU. In Transformers, those are generated on CPU, which slows down the process. This is also done in your Model References.

**Review by libinta — APPROVED** [review]
